quinn-proto/src/crypto/rustls.rs

```diff
@@ -106,33 +118,39 @@ impl crypto::Session for TlsSession {
         }
     }
 
-    fn write_handshake(&mut self, buf: &mut Vec<u8>) -> Option<Self::Keys> {
+    fn write_handshake(&mut self, buf: &mut Vec<u8>) -> Option<CryptoSpace<Self>> {
         let secrets = self.write_hs(buf)?;
+        let (local, remote) = select_secrets(&secrets, self.side());
         let suite = self
             .get_negotiated_ciphersuite()
             .expect("should not get secrets without cipher suite");
-        Some(Crypto::new(
-            self.side(),
-            suite.get_aead_alg(),
-            secrets.client,
-            secrets.server,
-        ))
-    }
 
-    fn update_keys(&self, keys: &Self::Keys) -> Self::Keys {
-        let (client_secret, server_secret) = match self.side() {
-            Side::Client => (&keys.local_secret, &keys.remote_secret),
-            Side::Server => (&keys.remote_secret, &keys.local_secret),
-        };
+        let aead = suite.get_aead_alg();
+        let result = Some(CryptoSpace {
```

Comment:
I feel like we could reduce coupling here by having the `ring` module contain the code to create a `CryptoSpace` from `aead`, `local` and `remote`?